### PR TITLE
fix(clang-plugin): close P1 gaps from latest-main plugin review (ADR-038 C.10-C.12)

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Implements the recommendations from the "Latest-main Clang plugin review" (reviewing commit `9c49163` / PR #548), which found the ADR-038 C.8 canonical fact-set work materially improved but flagged several P1/P2 correctness gaps. This PR closes the P1 items and the most impactful P2s; new ADR-038 sections (C.10-C.12) document what changed.

## Motivation

The review's core findings:
1. `source_edges` were collected by both producers but never reached the L5 graph (`SourceAbiSurface` had no edge field), so the separate `call_graph`/`type_graph` replay passes remained the only real edge source — defeating the plugin's zero-extra-parse goal.
2. The Clang JSON-AST replay resolved an overloaded callee's compact `referencedDecl` stub by bare name (verified: real Clang never puts `mangledName` on that stub), collapsing overloads/ctors/dtors onto one identity.
3. Incompatible opaque body/template hashes were flagged as unreliable (`SOURCE_FACT_COVERAGE_INCOMPLETE`) but then diffed anyway.
4. The plugin/wrapper's per-TU fact filename and manifest had no target/library isolation — two libraries sharing one `out=` directory could silently clobber each other's facts.
5. A compaction rollback edge case, a silently-dropped typedef diagnostic, and thin C.6/scan-flow test coverage of the newer channels.

## Changes

- **PR1 (source_edges → L5 graph):** `SourceAbiSurface.source_edges` (new field) + `source_link.link_source_abi()` folds/dedupes TU edges onto it; `source_graph.fold_source_edges()` consumes it inside `build_source_graph()`, using the same `decl://`/`type://` node-id scheme the replay passes use so edges reconcile instead of duplicating. `inline._build_inline_graph()` skips the separate `call_graph`/`type_graph` replay when a full-scope `source_edges` collection is already confirmed complete (narrowed scopes still run the replay).
- **PR1b (identity fix):** `call_graph.py`'s `parse_clang_ast_calls()` now builds an `id_index` from every full `FunctionDecl`/`CXXMethodDecl`/... node seen in the same walk, resolving a call site's compact stub through it before falling back to bare-name — mirroring `type_graph.py`'s existing fix. Verified end-to-end against a live Clang 18 compile of an `int f(int)`/`double f(double)` pair.
- **PR2 (hash gating):** `fact_set.FactCompatibility`/`check_fact_compatibility()` gate `INLINE_BODY_CHANGED`/`TEMPLATE_BODY_CHANGED` on producer/version comparability (existence/removal findings are never gated), with a `hash_recipe_id` override for producers proven equivalent via C.6.
- **PR3 (target isolation):** plugin fact filename + `tu_id` fold in the library identity; `ensureManifest()`/`init_inputs_pack()` validate an existing manifest against the current invocation (plugin: loud warning, never aborts the build; wrapper: hard error); `abicheck inputs validate` rejects multi-target packs and hard fact-set-recipe mismatches within one pack.
- **PR4:** `compact_inputs_pack()` rejects an `--output-filename` colliding with an unrecognized pre-existing file (instead of silently overwriting it, per the review's own recommended fix); a public typedef whose underlying type can't be determined now records a diagnostic folded into `types` coverage.
- **Test extension:** `conformance.py` (C.6) and `scan_flow.py` now cover `source_edges`/`read_files`/`fact_set`/`coverage`, including a planted overload-regression fixture, and assert the specific L5 graph edges are present (not just non-empty L4 entity counts).
- **ADR-038** gains C.10 (source_edges → L5 graph + identity fix), C.11 (hash gating), C.12 (target isolation + explicit note that the published 2.39× performance figure is *not* re-verified for the current collector — re-measuring requires an instrumented LLVM build this environment can't perform, so it's documented as an open gap rather than fabricated). **ADR-041**'s stale P1 roadmap item is updated to reflect completion.

## Testing

- Full fast lane: `pytest tests/ -m "not integration and not libabigail and not abicc and not slow and not golden"` — 14599 passed.
- `ruff check`, `mypy abicheck/`, `scripts/check_ai_readiness.py` — all clean (0 errors).
- The Python-side fixes (identity resolution, graph folding, hash gating, target-isolation validation) were also verified against **real Clang 18** compiles in this environment (live AST dumps, `abicheck-cc` wrapper runs, `abicheck dump`/`merge`/`compare` end-to-end) — not just unit tests against synthetic fixtures.
- The C++ plugin changes could not be compiled/linked in this sandbox (no full Clang C++ API headers available via apt — only `libclang-cpp` shared lib); the new code reuses the exact same LLVM API idioms (`MemoryBuffer::getFile`, `json::parse`+`consumeError`, `getAsObject`/`getString`) already compiling elsewhere in this file. The `clang-plugin` CI workflow will build and run the full plugin matrix (LLVM 16/17/18) against these changes.

## Checklist

- [x] Tests added / updated for new behaviour
- [x] `CHANGELOG.md` updated (under `[Unreleased]`)
- [x] Docs updated (ADR-038 C.10-C.12, ADR-041 roadmap)
- [x] `ruff`/`mypy` clean
- [ ] Plugin C++ build verified by CI (not buildable in this sandbox — see Testing)

---
_Generated by [Claude Code](https://claude.ai/code/session_01StWyhQn37w1gvG9uuWq86E)_

## Summary by CodeRabbit

* **New Features**
  * Source dependency and call relationships are now included more completely in generated ABI graphs.
  * Overloaded function calls are resolved to the correct callee.
  * Fact comparisons now account for producer and hashing compatibility.

* **Bug Fixes**
  * Prevented misleading inline/template body-change findings when hashes cannot be reliably compared.
  * Improved diagnostics for unavailable public typedef information.
  * Prevented cross-target file collisions and unsafe pack overwrites.

* **Validation**
  * Packs containing conflicting targets, libraries, versions, or fact contracts are now rejected with clear errors.

### CODE REVIEW

File: abicheck/buildsource/inline.py [hunk 1]
```diff
@@ -1598,22 +1621,34 @@ def _build_inline_graph(
...
+        # frontend invocation already produced complete source_edges for this
+        # same broad scope (ADR-038 C.9 / PR1) -- build_source_graph() above
+        # already folded them in via fold_source_edges. A narrowed scope
+        # (changed_paths / call_graph_units) is never skipped: narrowed
+        # source_edges only proves completeness for the TUs L4 actually
+        # parsed, same as extractor_pass_fully_covered's full/narrowed split.
+        broad_pass = not changed_paths and call_graph_units is None
+        if broad_pass and _source_edges_confirmed(surface):
+            graph.extractor_passes["call_graph"] = True
+            graph.extractor_passes["type_graph"] = True
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/inline.py]: Good catch, confirmed. `fold_source_edges`'s raw `source_edges` rows carry only bare endpoint identities, not the `dst_file`/project-file provenance `fold_call_graph`/`fold_type_graph` attach via `project_files`, which `is_internal_dependency_node`/`crosscheck.public_to_internal_dependency` need to classify an unannotated node as internal — skipping the replay could silently drop that detection.

Fixed in 0af1b90: reverted the skip, both replay passes now always run unconditionally when `with_call_graph` is set (as before this fix). `fold_source_edges` still runs and reconciles with the replay's edges via the same dedup key, so it remains a net addition, just not (yet) a substitute for the replay's provenance. Documented as a follow-up in ADR-038 C.10.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/source_diff.py [hunk 2]
```diff
@@ -239,15 +256,16 @@ def diff_source_abi(old: SourceAbiSurface, new: SourceAbiSurface) -> list[Change
     The result is an ordinary list of :class:`Change` objects ready to fold into
     a ``DiffResult`` and run through the existing verdict/policy pipeline.
     """
+    compat = check_fact_compatibility(_surface_fact_set(old), _surface_fact_set(new))
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/source_diff.py:276]: Confirmed the `source_diff.py` half of this (message accuracy) and fixed in 4f0e770: the suppression note now describes the actual fact-compatibility failure (name/version mismatch vs. recipe drift) instead of always blaming a missing `hash_recipe_id`, and separately notes when `source_edges` reconciliation is skipped.

On the broader ask (suppress/downgrade `_diff_generated`/`_diff_typedefs`/etc. when `structured_facts_comparable` is false): that's intentional, not an oversight — `check_fact_compatibility`'s own docstring and the original C.8 design say compiler-neutral structured facts (signatures, declarations, type shapes) should *always* be compared where valid; only a hard fact_set name/version mismatch (which already gates the whole comparison via `SOURCE_FACT_COVERAGE_INCOMPLETE`) invalidates them. `source_edges_comparable` gating the *L5 graph* structural diff (`source_graph_findings.diff_source_graph_findings`) is a real, separate gap — that function only receives two `SourceGraphSummary` objects today, not fact-set context, so threading `FactCompatibility` through it is a larger architectural change than fits this PR. Tracking as follow-up rather than bundling in here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/fact_set.py [hunk 3]
```diff
@@ -266,3 +300,93 @@ def check_fact_set_compatibility(
...
+#: hash_recipe_id -- a differential conformance run can prove two different
+#: producer identities emit byte-comparable hashes.
+_RECIPE_OVERRIDABLE_RULES = frozenset(
+    {"producer_mismatch", "producer_version_mismatch", "compiler_version_mismatch"}
+)
+
+#: source_edges endpoint identities are derived from the same producer/
+#: compiler-dependent AST-dump recipe as opaque hashes (mangled-name
+#: resolution, canonicalization), so they are gated by the same rule set.
+_SOURCE_EDGE_OVERRIDABLE_RULES = _RECIPE_OVERRIDABLE_RULES
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/fact_set.py:344]: Confirmed and fixed in 4f0e770, along your proposed diff: added `compiler_family_mismatch` to `_RECIPE_OVERRIDABLE_RULES` (also extends `_SOURCE_EDGE_OVERRIDABLE_RULES`, which aliases it) and included `compiler_family` in the fallback `hash_recipe_id` tuple. Added tests covering both the gating and the fallback-id-divergence behavior.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/inputs_validate.py [hunk 4]
```diff
@@ -87,6 +92,76 @@ def _duplicate_tu_ids(tus: list[SourceAbiTu]) -> list[str]:
...
+    two TUs in the *same* pack is stronger: it means the pack literally does
+    not agree on what mandatory-family contract it collected under, which
+    ``check_fact_set_compatibility``'s rule 1/2 already treats as an "error"
+    for an old/new comparison — the same severity applies just as much within
+    one pack (latest-main Clang plugin review, PR3).
+    """
+    distinct = list({tuple(sorted(tu.fact_set.items())) for tu in tus if tu.fact_set})
+    if len(distinct) < 2:
+        return []
+    fact_sets = [dict(d) for d in distinct]
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/inputs_validate.py]: Confirmed and fixed in 4f0e770, per your proposed approach (dict equality instead of hashing). Added a regression test with a nested-dict `fact_set` value to lock in the crash fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/source_abi.py [hunk 5]
```diff
@@ -577,6 +587,7 @@ def _ents(key: str) -> list[SourceEntity]:
             reachable_macros=_ents("macros"),
             reachable_templates=_ents("templates"),
             reachable_inline_bodies=_ents("inline_bodies"),
+            source_edges=list(d.get("source_edges", [])),
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/source_abi.py]: Confirmed and fixed in 4f0e770 via a shared `_edge_list()` helper applied to both `SourceAbiTu.from_dict` and `SourceAbiSurface.from_dict` (both had the same gap). Added tests for `None`/string/mixed-junk inputs on both classes.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/source_link.py [hunk 6]
```diff
@@ -1148,7 +1149,21 @@ def link_source_abi(
...
     for tu in tus:
+        for edge in tu.source_edges:
+            if not isinstance(edge, dict):
+                continue
+            edge_key = (
+                str(edge.get("edge", "")),
+                str(edge.get("src", "")),
+                str(edge.get("dst", "")),
+            )
+            if not all(edge_key) or edge_key in seen_edge_keys:
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/source_link.py]: Confirmed and fixed in 4f0e770 per your proposed approach (validate as real strings, don't stringify). Added a regression test with `None`/list/`None` identity fields.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: CHANGELOG.md [hunk 7]
```diff
@@ -432,6 +432,39 @@ Versioning: [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
...
+  `SourceAbiTu.source_edges` (call/type-relationship facts the plugin and the
+  reference `clang.py` extractor both collect during their existing frontend
+  pass) was serialized and round-tripped but `SourceAbiSurface` had no edge
+  field at all, so nothing ever folded them into the L5 graph — the separate
+  `call_graph`/`type_graph` Clang AST replay passes remained the only
+  producer of those edges end-to-end. `source_link.link_source_abi()` now
+  folds them onto a new `SourceAbiSurface.source_edges`, and
+  `source_graph.fold_source_edges()` consumes it from inside
+  `build_source_graph()`; the separate replay passes are skipped when a
+  full-scope `source_edges` collection is already confirmed complete.
```
[the remark this answers is not in the document]
Comment (napetrov) [reply CHANGELOG.md]: Fixed in 4f0e770 — corrected the release note to describe the actual (reverted-to-unconditional) replay behavior.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: contrib/abicheck-clang-plugin/AbicheckFactsPlugin.cpp [hunk 8]
```diff
@@ -3010,10 +3046,62 @@ class FactsConsumer : public ASTConsumer {
...
+    }
+  }
+
   void ensureManifest() {
     std::string manifestPath = OutDir + "/manifest.json";
-    if (llvm::sys::fs::exists(manifestPath))
+    if (llvm::sys::fs::exists(manifestPath)) {
+      checkManifestTargetAgreement(manifestPath);
       return;
+    }
```
[the remark this answers is not in the document]
Comment (napetrov) [reply contrib/abicheck-clang-plugin/AbicheckFactsPlugin.cpp]: Confirmed — real race, well-diagnosed. Fixed in 4f0e770: `ensureManifest` now atomically claims `manifest.json` itself via `CD_CreateNew` (fails with `file_exists` if another parallel compile already created or is creating it), instead of the `exists()`-then-tmp-file-rename TOCTOU. This makes "does a manifest exist" and "claim the right to create it" one indivisible syscall, matching the same exclusive-create primitive `claimFirstRootsWarning`/`claimFirstInferenceNote` already use elsewhere in this file for the identical once-across-parallel-compiles problem. Verified with a local clang++ 18 toolchain that the new code is consistent with the existing `openFile`/`CD_CreateNew` calling convention used by those two functions (couldn't do a full CMake link-test here — the environment lacks the static `libclangBasic.a`/component libs the reference CMakeLists.txt needs — but the real `clang-plugin` CI workflow builds and runs conformance/scan-flow against it).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: contrib/abicheck-clang-plugin/tests/conformance.py [hunk 9]
```diff
@@ -50,19 +74,173 @@
...
+            if not tu.fact_set:
+                continue
+            for family in FACT_FAMILIES:
+                state = tu.coverage.get(family)
+                if state in INCOMPLETE_COVERAGE_STATES:
+                    warnings.append(
+                        f"{label} TU {tu.tu_id} family {family!r} reported "
+                        f"{state!r} coverage on the C.6 fixture"
+                    )
+    return warnings
```
[the remark this answers is not in the document]
Comment (napetrov) [reply contrib/abicheck-clang-plugin/tests/conformance.py:250]: Confirmed the underlying gap, applied a corrected version in 4f0e770: used `state not in COVERAGE_STATES or state in ("partial", "failed")` rather than the literal suggested diff. The plain `{"complete", "empty-confirmed"}` allowlist would have broken the legitimate case where a family's documented state is `"unsupported"` (e.g. the plugin never collects `read_files` — see `FACT_FAMILIES`'s docstring) by turning every such run into a spurious warning. The real gap you found — `None`/an unrecognized string silently passing because neither is in `INCOMPLETE_COVERAGE_STATES` — is still closed by this condition.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/source_graph.py [hunk 10]
```diff
@@ -1082,6 +1082,94 @@ def header_declares(entity: SourceEntity, node_id: str, confidence: str) -> None
         )
         header_declares(ent, mid, conf)
 
+    fold_source_edges(graph, surface.source_edges)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/source_graph.py:1085]: Confirmed. Fixed in a967373: added `mark_source_edges_extractor_coverage()` in `source_graph.py`, applied at the one call site that never runs a replay — `inputs_pack.ingest_inputs_pack`. It translates a confirmed-complete/empty-confirmed `source_edges` rollup into `extractor_passes["call_graph"]`/`["type_graph"] = True`.

Note it's deliberately *not* applied inside `build_source_graph()` itself or in the two callers that always run the replay right after (`inline._build_inline_graph`, `cli_buildsource_helpers`) — those already track full-vs-narrowed coverage precisely via `extractor_pass_fully_covered()`/`narrowed_pass_confirmed()`, which a bare `source_edges` rollup can't distinguish (a "complete" rollup doesn't by itself tell you whether the run was scope-narrowed). A Flow-2 ingested pack has no narrowing concept at all — it always reflects whatever the build compiled — so marking full coverage there is safe.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/inputs_emit.py [hunk 11]
```diff
@@ -144,6 +166,22 @@ def init_inputs_pack(
                     f"{mpath} does not declare kind: {INPUTS_KIND} — not a "
                     "Flow-2 abicheck_inputs pack."
                 )
+            existing_library = str(data.get("library") or "")
+            if library and existing_library and library != existing_library:
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/inputs_emit.py]: Confirmed — same TOCTOU shape as the Clang plugin's `ensureManifest`, just on the Python wrapper side. Fixed in a967373: `init_inputs_pack` now claims `manifest.json` atomically via write-temp + `os.link()` (all-or-nothing — a losing reader always sees either nothing or a fully-written manifest, never a torn one, which a raw `O_CREAT|O_EXCL` direct write could risk). On `FileExistsError` it falls through to the existing read+validate path, so a losing call now correctly raises on a library/version conflict instead of silently losing the race. Added two tests simulating the race via a monkeypatched `os.link`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/inputs_pack.py [hunk 12]
```diff
@@ -619,9 +629,20 @@ def ingest_inputs_pack(
...
-        graph = build_source_graph(build_evidence or BuildEvidence(), source_abi=surface)
+        graph = build_source_graph(
+            build_evidence or BuildEvidence(), source_abi=surface
+        )
+        # No call/type-graph replay ever runs for an ingested Flow-2 pack (pure
+        # parsing, ADR-035 D5) -- translate a confirmed-complete source_edges
+        # rollup into the same extractor-pass coverage a replay would have
+        # recorded, or the decl-dependency crosscheck reads "never ran" and
+        # suppresses a real finding as a coverage artifact (Codex review).
+        mark_source_edges_extractor_coverage(graph, surface)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/inputs_pack.py:645]: Confirmed and fixed in fc93387: added `mark_source_edges_extractor_coverage(combined.source_graph, combined.source_abi)` right after the rebuild in `_relink_combined_against_exports()` (`cli_buildsource_merge.py`). Added a regression test (`test_dump_inputs_preserves_source_edges_coverage_across_export_relink`) exercising `embed_inputs_pack` end-to-end with a no-exported-symbols pack (which triggers the relink) and a complete `source_edges` rollup, asserting `extractor_passes` survives.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/source_graph.py [hunk 13]
```diff
@@ -1082,6 +1082,130 @@ def header_declares(entity: SourceEntity, node_id: str, confidence: str) -> None
...
+        ):
+            if not graph.has_node(node_id):
+                graph.add_node(
+                    GraphNode(
+                        id=node_id,
+                        kind=node_kind,
+                        label=ident,
+                        provenance=provenance,
+                        confidence=confidence,
+                    )
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/source_graph.py:1155]: Confirmed — this is real, and it's deeper than the pass-coverage gate my last fix (`mark_source_edges_extractor_coverage`) closed. That fix only makes `_dependency_kinds_covered`/`_common_dependency_edge_kinds` stop treating a Flow-2-ingested graph's zero-edge families as "never examined." It does *not* touch `is_internal_dependency_node()`, which is a separate, per-node gate that reads the node's own `visibility`/`defined_in_project` attrs (or `decl_to_file`, which only covers *public* reachable declarations) — completely independent of `extractor_passes`. A node created purely by `fold_source_edges` never gets either, so a genuinely private target still can't classify as internal, and `PUBLIC_API_INTERNAL_DEPENDENCY_ADDED` still can't fire for it, coverage gate notwithstanding.

Both gates need to pass, and only one is fixed. Properly closing this requires the *producer* side (the plugin's C++ wire format and `clang_source_edges.build_source_edges()`) to actually emit `dst_file`/project-file provenance per edge — the same follow-up ADR-038 C.10 already flags for the inline-scan optimization, but here it's not an optimization risk, it's the *only* path for a Flow-2-only pack (no compile DB, no replay possible at all). That's a real cross-cutting change (wire schema + both producers), too large to bundle into this PR as a quick patch, so I've documented it explicitly in ADR-038 C.10 (pushed in 25adb33) as a known, accepted limitation of the ingestion-only path rather than attempting an incomplete fix here. Leaving this thread open since it isn't actually resolved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: contrib/abicheck-clang-plugin/AbicheckFactsPlugin.cpp [hunk 14]
```diff
@@ -3036,17 +3121,30 @@ class FactsConsumer : public ASTConsumer {
...
+    // claimFirstInferenceNote already use for this file's other
+    // once-across-parallel-compiles claims.
     int fd = -1;
-    if (llvm::sys::fs::createUniqueFile(
-            llvm::Twine(OutDir) + "/.manifest.%%%%%%.tmp", fd, tmp))
+    std::error_code ec = llvm::sys::fs::openFile(
+        manifestPath, fd, llvm::sys::fs::CD_CreateNew, llvm::sys::fs::FA_Write,
+        llvm::sys::fs::OF_None);
+    if (ec == std::errc::file_exists) {
+      checkManifestTargetAgreement(manifestPath);
```
[the remark this answers is not in the document]
Comment (napetrov) [reply contrib/abicheck-clang-plugin/AbicheckFactsPlugin.cpp]: Sharp catch — confirmed, and you're right that this defeats the previous fix in exactly the way it was meant to prevent. Fixed in 25adb33: `ensureManifest` now writes the full manifest to a private temp file first, then publishes it via `create_hard_link()` (verified against real LLVM 18 `sys::fs` headers — `create_hard_link(to, from)` creates `from` pointing at `to`'s content, failing with `file_exists` if `from` already exists, without ever creating a partial file there). `manifestPath` is now only ever "absent" or "fully written" — never a torn intermediate state — so `checkManifestTargetAgreement()` always reads complete JSON. Same idiom as the Python wrapper's `os.link()`-based fix for the equivalent race in `inputs_emit.init_inputs_pack`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Re Codex's "Re-finalize after marking source-edge coverage" finding on `cli_buildsource_merge.py:256-258` (posted as a general comment, not an inline thread): confirmed and fixed in b13df8f. `build_source_graph()` already calls `finalize()` once before returning, computing `graph.coverage`'s `call_edges`/`type_edges`/`reference_edges` "collected" flags from `extractor_passes` as of that moment. `mark_source_edges_extractor_coverage()` mutates `extractor_passes` afterward, so `finalize()` needed to re-run — otherwise the serialized `coverage` summary still said "collected: false" even though the pass flags were now true, exactly as reported. `ingest_inputs_pack()` already did this correctly; `_relink_combined_against_exports()` didn't. Strengthened the existing regression test to assert on `graph.coverage` directly (confirmed it fails without the fix).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/source_diff.py [hunk 15]
```diff
@@ -239,15 +263,16 @@ def diff_source_abi(old: SourceAbiSurface, new: SourceAbiSurface) -> list[Change
...
     changes: list[Change] = []
-    changes.extend(_diff_fact_coverage(old, new))
+    changes.extend(_diff_fact_coverage(old, new, compat))
     changes.extend(_diff_generated(old, new))
     changes.extend(_diff_typedefs(old, new))
     changes.extend(_diff_macros(old, new))
     changes.extend(_diff_concepts(old, new))
     changes.extend(_diff_declarations(old, new))
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/source_diff.py]: Confirmed and fixed in d1886aa (caught independently during a full self-review of the PR, then this landed too — same finding). Gated the three *removal*-detection loops (`GENERATED_HEADER_CHANGED`'s removal half, `PUBLIC_TYPEDEF_REMOVED`, `PUBLIC_MACRO_REMOVED`) on `compat.structured_facts_comparable`.

Narrower than "gate the structured passes" though: content comparisons for identities present on **both** sides (`PUBLIC_MACRO_VALUE_CHANGED`, `PUBLIC_TYPEDEF_TARGET_CHANGED`, default-argument/constexpr changes in `_diff_declarations`) are intentionally left ungated — a type/signature hash means the same thing regardless of which mandatory-family contract collected it. What a fact_set name/version mismatch actually invalidates is *existence* claims: an entity's absence might just mean the old contract never mandated collecting that family, not that it was removed — the same reasoning `_surface_has_facts(new)` already applies for the "no facts at all" case, just now extended to the "incompatible contracts" case too. `_diff_declarations` has no removal path so needed no change; `_diff_mappings`'s "identity gone" check is anchored to authoritative L0 binary-export evidence (a stronger, different signal), so it's left as-is.

Also refined `FactCompatibility.structured_facts_comparable`'s docstring to say precisely what it gates, since the previous wording didn't specify and is exactly what let this go unconsumed.

---
_Generated by [Claude Code](https://claude.ai/code)_
